src/ec/Jacobi.mo

```diff
@@ -0,0 +1,420 @@
+      case (#infinity (_), _) {
+        point2
+      };
+      case (_, #infinity (_)) {
+        point1
+      };
+      case (#point (X1, Y1, Z1, curve), #point (X2, Y2, Z2, _)) {
+        let (X3, Y3, Z3) = _add(X1.value, Y1.value, Z1.value,
+          X2.value, Y2.value, Z2.value, curve.a, curve.p);
+        if (Y3 == 0 or Z3 == 0) {
+          #infinity (curve);
+        } else {
+          #point (fpFromInt(X3, curve), fpFromInt(Y3, curve),
+            fpFromInt(Z3, curve), curve);
+        };
+      };
+    };
+  };
+
+  func doDouble(X1 : Int, Y1 : Int, Z1 : Int, a : Nat, p : Nat) : (Int, Int, Int) {
+   if (Z1 == 1) {
+     return doubleWithZ1(X1, Y1, a, p)
+   };
+   if (Y1 == 0 or Z1 == 0) {
+     return (0, 0, 0);
+   };
+
+   let (XX, YY) = (X1 * X1 % p, Y1 * Y1 % p);
+   if (YY == 0) {
+     return (0, 0, 0);
+   };
+
+   let YYYY : Int = YY * YY % p;
+   let ZZ : Int = Z1 * Z1 % p;
+   let S : Int = 2 * ((X1 + YY) ** 2 - XX - YYYY) % p;
+   let M : Int = (3 * XX + a * ZZ * ZZ) % p;
+   let T : Int = (M * M - 2 * S) % p;
+
+   let Y3 = (M * (S - T) - 8 * YYYY) % p;
+   let Z3 = ((Y1 + Z1) ** 2 - YY - ZZ) % p;
+
+   return (T, Y3, Z3);
+  };
+
+  func doubleWithZ1(X1: Int, Y1: Int, a : Nat, p : Nat): (Int, Int, Int) {
+      let (XX, YY) = (X1 * X1 % p, Y1 * Y1 % p);
+
+      if(YY == 0) {
+        return (0, 0, 0);
+      };
+
+      let YYYY : Int = YY * YY % p;
+      let S : Int = 2 * ((X1 + YY) ** 2 - XX - YYYY) % p;
+      let M : Int = 3 * XX + a;
+      let T : Int = (M * M - 2 * S) % p;
+      let Y3 : Int = (M * (S - T) - 8 * YYYY) % p;
+      let Z3 : Int = 2 * Y1 % p;
+      return (T, Y3, Z3);
+  };
+
+  func addWithEqZ(X1 : Int, Y1 : Int, Z1 : Int,
+    X2 : Int, Y2 : Int, a : Nat, p : Nat) : (Int, Int, Int) {
+
+    let A : Int = (X2 - X1) ** 2 % p;
+    let B : Int = X1 * A % p;
+    let C : Int = X2 * A;
+    let D : Int = (Y2 - Y1) ** 2 % p;
+
+    if (A == 0 and D == 0) {
+      return doDouble(X1, Y1, Z1, a, p);
+    };
+
+    let X3 : Int = (D - B - C) % p;
+    let Y3 : Int = ((Y2 - Y1) * (B - X3) - Y1 * (C - B)) % p;
+    let Z3 : Int = Z1 * (X2 - X1) % p;
+
+    return (X3, Y3, Z3);
+  };
+
+  func addWithZ1(X1 : Int, Y1 : Int, X2 : Int, Y2 : Int,
+    a : Nat, p : Nat) : (Int, Int, Int) {
+    let H : Int = X2 - X1;
+    let HH : Int = H * H;
+    let I : Int = 4 * HH % p;
+    let J : Int = H * I;
+    let r : Int = 2 * (Y2 - Y1);
+
+    if(H == 0 and r == 0) {
+      return doubleWithZ1(X1, Y1, a, p);
+    };
+
+    let V : Int = X1 * I;
+    let X3 : Int = (r**2 - J - 2 * V) % p;
+    let Y3 : Int = (r * (V - X3) - 2 * Y1 * J) % p;
+    let Z3 : Int = 2 * H % p;
+
+    return (X3, Y3, Z3);
+  };
+
+  func addWithZ2Eq1(X1 : Int, Y1 : Int, Z1 : Int, X2 : Int, Y2 : Int,
+    a : Nat, p : Nat) : (Int, Int, Int) {
+
+    let Z1Z1 : Int = Z1 * Z1 % p;
+    let (U2, S2) = (X2 * Z1Z1 % p, Y2 * Z1 * Z1Z1 % p);
+    let H : Int = (U2 - X1) % p;
+    let HH : Int = H * H % p;
+    let I : Int = 4 * HH % p;
```

Comment:
It would be useful to run all of these functions through a wrapper (eg from `Field`) which handles the reductions. Then if you implement something like Barrett you just have to modify the one function.

Also, having an explicit function handling the squaring case is both a bit more readable and allows using the faster squaring algorithms (squaring is `O(n**1.5)` vs multiplication which is `O(n**2)`)